### PR TITLE
[feat][txn]Implement transactionCoordinatorClient

### PR DESCRIPTION
**Pull request by liangyepianzhou**
Master Issue:https://github.com/apache/pulsar-client-go/issues/932
### Motivation
Implement transaction coordinator client.
### Modifications
1. Implement transaction coordinator
2. implement transaction API
3. Add metric and test
### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar/transaction_test.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	pb "github.com/apache/pulsar-client-go/pulsar/internal/pulsar_proto"
+	"github.com/stretchr/testify/assert"
+
+	"testing"
+	"time"
+)
+
+func TestTCClient(t *testing.T) {
+	//1. Prepare: create PulsarClient and init transaction coordinator client.
+	topic := "my-topic"
```
Comment (shibd) [inline pulsar/transaction_test.go:30]:
```suggestion
	topic := newTopicName()
```

File: pulsar/transaction_test.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	defer tc.close()
+	defer client.Close()
+}
+
+/*
+Create a transaction coordinator client to send request
+*/
+func createTcClient(t *testing.T) (*transactionCoordinatorClient, *client) {
+	c, err := NewClient(ClientOptions{
+		URL:                   webServiceURLTLS,
```
Comment (shibd) [inline pulsar/transaction_test.go:87]: Does Transaction have to have TLS enabled?

File: Makefile [hunk 3]
```diff
@@ -18,7 +18,7 @@
 #
 
 IMAGE_NAME = pulsar-client-go-test:latest
-PULSAR_VERSION ?= 2.8.3
+PULSAR_VERSION ?= 2.10.3
```
[the remark this answers is not in the document]
Comment (liangyepianzhou) [reply Makefile:21]: CI failed again after I changed it to `latest`, so I can only change it to 2.10.3. The specific reason is not clear.

File: pulsar/transaction_test.go (hunk 2 above)
Comment (liangyepianzhou) [reply pulsar/transaction_test.go:87]: The transaction needs to get a connection from "pulsar/system/transaction_coordinator_assign".
And the client only has the authority of the `public` tenant in default.
So we need to add the authentication here.

File: pulsar/transactionCoordinatorClient.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (nodece) [inline pulsar/transactionCoordinatorClient.go:1]: `pulsar/transactionCoordinatorClient.go` rename to `pulsar/transaction_coordinator_client.go`

File: pulsar/transactionCoordinatorClient.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+	for _, con := range tc.cons {
+		con.Close()
+	}
+}
+
+/*
+*
+New a transactionImpl which can be used to guarantee exactly-once semantics.
+*/
+func (tc *transactionCoordinatorClient) newTransaction(timeout time.Duration) (TxnID, error) {
```
Comment (nodece) [inline pulsar/transactionCoordinatorClient.go:115]: It is better `*TxnID, error`

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (shibd) [reply Makefile:21]: > CI failed again after I changed it to latest, so I can only change it to 2.10.3. The specific reason is not clear.
> Fixed version is right.

We can fix it in other PRs. But, Do we need the Go client to be compatible with the latest major version (the current is 2.11.0)?

File: pulsar/transaction_coordinator_client.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+/*
+*
+New a transactionImpl which can be used to guarantee exactly-once semantics.
+*/
+func (tc *transactionCoordinatorClient) newTransaction(timeout time.Duration) (*TxnID, error) {
+	err := tc.canSendRequest()
+	if err != nil {
+		return &TxnID{}, err
```
Comment (nodece) [inline pulsar/transaction_coordinator_client.go:118]:
```suggestion
		return nil, err
```

File: pulsar/transaction_coordinator_client.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	nextTcID := tc.nextTCNumber()
+	cmdNewTxn := &pb.CommandNewTxn{
+		RequestId:     proto.Uint64(requestID),
+		TcId:          proto.Uint64(nextTcID),
+		TxnTtlSeconds: proto.Uint64(uint64(timeout.Milliseconds())),
+	}
+
+	cnx, err := tc.client.rpcClient.RequestOnCnx(tc.cons[nextTcID], requestID, pb.BaseCommand_NEW_TXN, cmdNewTxn)
+	if err != nil {
+		return &TxnID{}, err
```
Comment (nodece) [inline pulsar/transaction_coordinator_client.go:130]:
```suggestion
		return nil, err
```

File: pulsar/transaction_coordinator_client.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+	_, err = tc.client.rpcClient.RequestOnCnx(tc.cons[id.mostSigBits], requestID,
+		pb.BaseCommand_ADD_SUBSCRIPTION_TO_TXN, cmdAddSubscription)
+	return err
+}
+
+/*
+*
+Commit or abort the transactionImpl.
+*/
+func (tc *transactionCoordinatorClient) endTxn(id *TxnID, action pb.TxnAction) error {
```
Comment (nodece) [inline pulsar/transaction_coordinator_client.go:188]:
```suggestion
func (tc *transactionCoordinatorClient) endTxn(id TxnID, action pb.TxnAction) error {
```

File: pulsar/transaction_coordinator_client.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+	_, err = tc.client.rpcClient.RequestOnCnx(tc.cons[id.mostSigBits], requestID,
+		pb.BaseCommand_ADD_PARTITION_TO_TXN, cmdAddPartitions)
+	return err
+}
+
+/*
+Register the subscription which acked messages with the transactionImpl.
+And this can be used when ending the transactionImpl.
+*/
+func (tc *transactionCoordinatorClient) addSubscriptionToTxn(id *TxnID, topic string, subscription string) error {
```
Comment (nodece) [inline pulsar/transaction_coordinator_client.go:163]:
```suggestion
func (tc *transactionCoordinatorClient) addSubscriptionToTxn(id TxnID, topic string, subscription string) error {
```

File: pulsar/transaction_coordinator_client.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+	return &TxnID{*cnx.Response.NewTxnResponse.TxnidMostBits,
+		*cnx.Response.NewTxnResponse.TxnidLeastBits}, nil
+}
+
+/*
+*
+Register the partitions which published messages with the transactionImpl.
+And this can be used when ending the transactionImpl.
+*/
+func (tc *transactionCoordinatorClient) addPublishPartitionToTxn(id *TxnID, partitions []string) error {
```
Comment (nodece) [inline pulsar/transaction_coordinator_client.go:142]:
```suggestion
func (tc *transactionCoordinatorClient) addPublishPartitionToTxn(id TxnID, partitions []string) error {
```

File: pulsar/helper_for_test.go [hunk 11]
```diff
@@ -157,6 +157,13 @@ func topicStats(topic string) (map[string]interface{}, error) {
 	return metadata, err
 }
 
+func transactionStats(id *TxnID) (map[string]interface{}, error) {
```
Comment (nodece) [inline pulsar/helper_for_test.go:160]:
```suggestion
func transactionStats(id TxnID) (map[string]interface{}, error) {
```

File: pulsar/transaction_coordinator_client.go (hunk 9 above)
Comment (liangyepianzhou) [reply pulsar/transaction_coordinator_client.go:163]: Why should we use TxnID instead of * TxnID here? I understand that using `*TxnID` can reduce the memory copy during value transfer.

File: pulsar/transaction_coordinator_client.go (hunk 8 above)
Comment (liangyepianzhou) [reply pulsar/transaction_coordinator_client.go:188]: Why should we use TxnID instead of * TxnID here? I understand that using *TxnID can reduce the memory copy during value transfer.

Comment (nodece) [reply pulsar/transaction_coordinator_client.go:188]: Sounds good. But when using the `*TxnID`, I always want to check if it is `nil`, so I suggest you use `TxnID`, this mean is not `nil`.

Comment (liangyepianzhou) [reply pulsar/transaction_coordinator_client.go:188]: This is the logic of internal calls, not the interface exposed to users. So it is not necessary to determine whether it is `nil` here.

Comment (nodece) [reply pulsar/transaction_coordinator_client.go:188]: Ok.

**Review by nodece — APPROVED** [review]
LGTM

File: pulsar/transaction_coordinator_client.go [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+
+	return &TxnID{*cnx.Response.NewTxnResponse.TxnidMostBits,
+		*cnx.Response.NewTxnResponse.TxnidLeastBits}, nil
+}
+
+/*
+*
+Register the partitions which published messages with the transactionImpl.
+And this can be used when ending the transactionImpl.
+*/
```
Comment (BewareMyPower) [inline pulsar/transaction_coordinator_client.go:141]: Please make the comment style consistent. Use 

```go
// line 1...
// line 2...
```

or

```go
/**
 * line1
 * line2
 * ...
 */
```

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: integration-tests/conf/standalone.conf [hunk 13]
```diff
@@ -111,8 +110,10 @@ superUserRoles=localhost,superUser,admin
 
 # Authentication settings of the broker itself. Used when the broker connects to other brokers,
 # either in same or other clusters
-brokerClientAuthenticationPlugin=
-brokerClientAuthenticationParameters=
+brokerClientTlsEnabled=true
+brokerClientTrustCertsFilePath=/pulsar/certs/cacert.pem
+brokerClientAuthenticationPlugin=org.apache.pulsar.client.impl.auth.AuthenticationTls
+brokerClientAuthenticationParameters={"tlsCertFile":"/pulsar/certs/client-cert.pem","tlsKeyFile":"/pulsar/certs/client-key.pem"}
```
Comment (BewareMyPower) [inline integration-tests/conf/standalone.conf:116]: Why did you add the TLS authentication? Does transaction depend on the TLS?

Comment (liangyepianzhou) [reply integration-tests/conf/standalone.conf:116]: Transaction coordinators use `brokerClient` to send commands to TB and TP.
So we need to add  TLS authentication to `brokerClient` here.

File: pulsar/client.go [hunk 14]
```diff
@@ -142,6 +142,8 @@ type ClientOptions struct {
 	// Specify metric registerer used to register metrics.
 	// Default prometheus.DefaultRegisterer
 	MetricsRegisterer prometheus.Registerer
+
+	IsEnableTransaction bool
```
Comment (BewareMyPower) [inline pulsar/client.go:146]: Could you remove the `Is` prefix? See other options that enable some feature:
- producer.go: EnableChunking
- consumer.go: EnableBatchIndexAcknowledgment, RetryEnable, EnableDefaultNackBackoffPolicy

File: pulsar/transaction_coordinator_client.go [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+	"github.com/apache/pulsar-client-go/pulsar/log"
+	"google.golang.org/protobuf/proto"
+)
+
+type transactionCoordinatorClient struct {
+	client                    *client
+	cons                      []internal.Connection
+	epoch                     uint64
+	semaphore                 internal.Semaphore
+	blockIfReachMaxPendingOps bool
```
Comment (BewareMyPower) [inline pulsar/transaction_coordinator_client.go:37]: Could this field be false? I see `transactionCoordinatorClient` should only be created by `newTransactionCoordinatorClientImpl`, which set it true. If it's always true, please remove this field.

File: pulsar/transaction_test.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+		assert.Equal(t, err.Error(), "http error status code: 404")
+	}
+	defer consumer.Close()
+	defer tc.close()
+	defer client.Close()
+}
+
+/*
+Create a transaction coordinator client to send request
+*/
```
Comment (BewareMyPower) [inline pulsar/transaction_test.go:84]:
```suggestion
// Create a transaction coordinator client to send request
```

File: pulsar/transaction_coordinator_client.go [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+	//The number of transactionImpl coordinators
+	tcNum uint64
```
Comment (BewareMyPower) [inline pulsar/transaction_coordinator_client.go:39]: This field is redundant. We can replace the use of it with `len(cons)`.

File: pulsar/transaction_coordinator_client.go [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+		con.Close()
+	}
+}
+
+// newTransaction new a transactionImpl which can be used to guarantee exactly-once semantics.
+func (tc *transactionCoordinatorClient) newTransaction(timeout time.Duration) (*TxnID, error) {
+	err := tc.canSendRequest()
+	if err != nil {
+		return nil, err
+	}
```
Comment (BewareMyPower) [inline pulsar/transaction_coordinator_client.go:114]:
```suggestion
        if err := tc.canSendRequest(); err != nil {
                return nil, err
        }
```

Simplify the code

File: pulsar/transaction_coordinator_client.go [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+		*cnx.Response.NewTxnResponse.TxnidLeastBits}, nil
+}
+
+// addPublishPartitionToTxn register the partitions which published messages with the transactionImpl.
+// And this can be used when ending the transactionImpl.
+func (tc *transactionCoordinatorClient) addPublishPartitionToTxn(id *TxnID, partitions []string) error {
+	err := tc.canSendRequest()
+	if err != nil {
+		return err
+	}
```
Comment (BewareMyPower) [inline pulsar/transaction_coordinator_client.go:138]:
```suggestion
        if err := tc.canSendRequest(); err != nil {
                return nil, err
        }
```

Simplify the code. The following `_, err = ...` in line 146 should become `_, err := ...` after this change.

File: pulsar/transaction_coordinator_client.go [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+	pb "github.com/apache/pulsar-client-go/pulsar/internal/pulsar_proto"
+	"github.com/apache/pulsar-client-go/pulsar/log"
+	"google.golang.org/protobuf/proto"
+)
+
+type transactionCoordinatorClient struct {
+	client                    *client
+	cons                      []internal.Connection
+	epoch                     uint64
+	semaphore                 internal.Semaphore
```
Comment (BewareMyPower) [inline pulsar/transaction_coordinator_client.go:36]: It's never released. You should release the semaphore for each response received

File: pulsar/client_impl_test.go [hunk 21]
```diff
@@ -425,7 +426,7 @@ func TestNamespaceTopics(t *testing.T) {
 		t.Fatal(err)
 	}
 	topic2 := fmt.Sprintf("%s/topic-2", namespace)
-	if err := httpPut("admin/v2/persistent/"+topic2, namespace); err != nil {
```
Comment (BewareMyPower) [inline pulsar/client_impl_test.go:428]: Could you explain changes in this test?

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: pulsar/client_impl_test.go (hunk 21 above)
Comment (liangyepianzhou) [reply pulsar/client_impl_test.go:428]: This request got a 400 response after enabling TLS.
I confirmed to @freeznet  it is an error code.
So I change its logic to be like the logic of topic 1.

**Review by BewareMyPower — APPROVED** [review]
